File: webui/src/modules/message/models/reaction.ts
import type { User } from '@/modules/auth/models/user'
import type { Message } from './message'

export class Reaction {
  public reactionId: number
  public messageId: number
  public userId: number
  public reaction: Blob
  public user?: User
  public message?: Message

  constructor(data: ReactionDTO) {
    this.reactionId = data.reaction_id
    this.messageId = data.message_id
    this.userId = data.user_id
    this.reaction = data.reaction
  }

  static fromJSON(json: ReactionDTO): Reaction {
    return new Reaction(json)
  }

  toJSON(): ReactionDTO {
    return {
      reaction_id: this.reactionId,
      message_id: this.messageId,
      user_id: this.userId,
      reaction: this.reaction,
    }
  }
}

export interface ReactionDTO {
  reaction_id: number
  message_id: number
  user_id: number
  reaction: Blob
}

File: service/shared/models/reaction.go
package models

import (
	"encoding/json"
)

type Reaction struct {
    ID        int    `json:"-"`
    MessageID int    `json:"-"`
    UserID    int    `json:"-"`
    Reaction  string `json:"-"`
    User      User   `json:"user"`
}

func (r Reaction) MarshalJSON() ([]byte, error) {
    return json.Marshal(&struct {
        ID        int    `json:"reaction_id"`
        MessageID int    `json:"message_id"`
        UserID    int    `json:"-"`
        Reaction  string `json:"reaction"`
        User      User   `json:"user"`
    }{
        ID:        r.ID,
        MessageID: r.MessageID,
        UserID:    r.UserID,
        Reaction:  r.Reaction,
        User:      r.User,
    })
}